content/src/content/docs/learn/tutorials/services-and-config.mdx

```diff
@@ -0,0 +1,381 @@
+---
+type: Tutorial
+title: Services and Config
+tags:
+  - some
+  - tags
+sidebar:
+  order: 5
+---
+
+import { Aside } from "@astrojs/starlight/components"
+
+In the previous sections we've seen how to write endpoints in Effect that call
+out to databases and other APIs. 
+
+In this section we're going to revisit that code and write it more
+"Effectfully." We're going to create services for our database and external API
+calls, and use Effect's `Config` module to pass configuration to our
+application. This will align our code more with the Effect best practices.
+
+We're going to start out creating a service for our database. As a reminder,
+here's what our `/users` endpoint looked like:
+
+```ts twoslash showLineNumbers=false
+import { HttpServerResponse } from "@effect/platform"
+import { Effect } from "effect"
+import { Client } from "pg"
+
+const db = new Client({ user: "postgres" })
+//---cut---
+const users = Effect.promise(async () => {
+  const { rows } = await db.query("SELECT * FROM users")
+  return { users: rows }
+}).pipe(
+  Effect.flatMap((body) => HttpServerResponse.json(body))
+)
+```
+
+## Defining a Database service
+
+The first thing we need is a `Tag` for this service, which looks like this:
+
+```ts twoslash showLineNumbers=false
+import { Context, Effect } from "effect";
+//---cut---
+class Database extends Context.Tag("Database")<
+  Database,
+  {
+    getUsers: () => Effect.Effect<any[], never, never>;
+  }
+>() {}
+```
+
+And then we can use the `Database.of` method to create an instance of the service:
+
+```ts twoslash showLineNumbers=false
+import { Context, Effect } from "effect";
+import { Client } from "pg"
+const db = new Client({ user: "postgres" })
+class Database extends Context.Tag("Database")<
+  Database,
+  {
+    getUsers: () => Effect.Effect<any[], never, never>;
+  }
+>() {}
+//---cut---
+const DatabaseLive = Database.of({
+    getUsers: () => Effect.promise(async () => {
+        const { rows } = await db.query("SELECT * FROM users")
+        return rows
+    }),
+})
+```
+
+The choice of `DatabaseLive` for the name is a common convention you will see
+in Effect code, where "live" is referring to the fact that this is the version
+of this service we will use in live, production code.
+
+When originally implementing the `/users` endpoint, I glossed over the fact that
+database calls can fail. I'm going to right that wrong here by using
+`Effect.tryPromise` and wrapping any error we get back from `pg`.
+
+```ts twoslash ins={1-3,8,15-19} showLineNumbers=false
+import { Context, Effect } from "effect";
+import { Client } from "pg"
+const db = new Client({ user: "postgres" })
+//---cut---
+class DatabaseError extends Error {
+  readonly _tag = "DatabaseError";
+}
```

Comment:
It would be worth adding a `constructor` here to take the `cause`:
```ts
constructor(readonly cause: unknown) {}
```

Reply:
JavaScript's default `Error` class already accepts a cause as part of its second argument:

```typescript
class DatabaseError extends Error {
  readonly _tag = "DatabaseError";
}

new DatabaseError("oh no", { cause: new Error("cause") });
```

Reply:
Wow! TIL! 👍

Reply:
Wow! TIL! 👍